primitives/utility/src/lib.rs

```diff
@@ -51,3 +58,212 @@ impl<T: UpwardMessageSender, W: WrapVersion> SendXcm for ParentAsUmp<T, W> {
 		}
 	}
 }
+
+/// Contains information to handle refund/payment for xcm-execution
+#[derive(Clone, Eq, PartialEq, Debug)]
+struct AssetTraderRefunder {
+	// The amount of weight bought minus the weigh already refunded
+	weight_outstanding: Weight,
+	// The concrete asset containing the asset location and outstanding balance
+	outstanding_concrete_asset: MultiAsset,
+}
+
+/// Charges for exercution in the first multiasset of those selected for fee payment
+/// Only succeeds for Concrete Fungible Assets
+/// First tries to convert the this MultiAsset into a local assetId
+/// Then charges for this assetId as described by FeeCharger
+/// Weight, paid balance, local asset Id and the multilocation is stored for
+/// later refund purposes
+/// Important: Errors if the Trader is being called twice by 2 BuyExecution instructions
+/// Alternatively we could just return payment in the aforementioned case
+pub struct TakeFirstAssetTrader<
+	AccountId,
+	FeeCharger: ChargeWeightInFungibles<AccountId, ConcreteAssets>,
+	Matcher: MatchesFungibles<ConcreteAssets::AssetId, ConcreteAssets::Balance>,
+	ConcreteAssets: fungibles::Mutate<AccountId> + fungibles::Transfer<AccountId> + fungibles::Balanced<AccountId>,
+	HandleRefund: TakeRevenue,
+>(
+	Option<AssetTraderRefunder>,
+	PhantomData<(AccountId, FeeCharger, Matcher, ConcreteAssets, HandleRefund)>,
+);
+impl<
+		AccountId,
+		FeeCharger: ChargeWeightInFungibles<AccountId, ConcreteAssets>,
+		Matcher: MatchesFungibles<ConcreteAssets::AssetId, ConcreteAssets::Balance>,
+		ConcreteAssets: fungibles::Mutate<AccountId>
+			+ fungibles::Transfer<AccountId>
+			+ fungibles::Balanced<AccountId>,
+		HandleRefund: TakeRevenue,
+	> WeightTrader
+	for TakeFirstAssetTrader<AccountId, FeeCharger, Matcher, ConcreteAssets, HandleRefund>
+{
+	fn new() -> Self {
+		Self(None, PhantomData)
+	}
+	// We take first multiasset
+	// Check whether we can convert fee to asset_fee (is_sufficient, min_deposit)
+	// If everything goes well, we charge.
+	fn buy_weight(
+		&mut self,
+		weight: Weight,
+		payment: xcm_executor::Assets,
```

Comment:
Still trying to understand this, the `buy_weight` function takes ownership of `payment` and only takes the first `MultiAsset` out of it, and the rest is then dropped after this function ends? That doesn't sound right to me.

Reply:
The rest is not dropped, but rather returned (the unused part). Whatever returns from here is credited back to the holding register.

Reply:
Where's the code for that? Is that from the `Drop` implementation of `xcm_executor::Assets`?

Reply:
Its part of the xcm-executor code for BuyExecution. https://github.com/paritytech/polkadot/blob/7e11707a55fcad806c71179c8cab6505cde65991/xcm/xcm-executor/src/lib.rs#L453.

Unless I am missing something, whatever we return from buy_weight should be credited to the holding